File: summary.jl
using CSV, DataFrames, Statistics
using Random: seed!
seed!(90125)

# Load the train set
path = joinpath(ENV["VOLTAIRE"], "data", "datasets", "SILFIAC")
data = CSV.read(joinpath(path, "train.csv"), DataFrame)

# Separate data from labels
labels = 0:23 .|> x -> "power_$(x)"
X = data[:, Not(vcat(Symbol.(labels), :date_col))]
y = data[:, labels] |> eachrow .|> mean

using Distributed
n_cpus = 4
workers = addprocs(n_cpus, exeflags="--project=.")

@everywhere using DataFrames, Dates, MLJ, MLJBase
@everywhere smape(ytrue, ypred)= 200 * mean(@. abs(ytrue - ypred) / (abs(ytrue) + abs(ypred)))
@everywhere PCA = @load PCA
@everywhere model = (@load EpsilonSVR)()
@everywhere pipe = @pipeline(
    FeatureSelector(features = name -> occursin("48.2_-3.1", string(name)),
                        ignore=true),
    Standardizer(),                 
    PCA(pratio=0.95),
    model,
    target=x->log.(x.+1),
    inverse=x->exp.(x).-1,
    name = "my_pipeline")

@everywhere using Random
@everywhere import Base.rand
@everywhere struct LogUniform
    a
    b    
end

@everywhere Base.rand(rng::Random.AbstractRNG, d::LogUniform) = exp(log(d.a) + (log(d.b) - log(d.a)) * rand(rng))


r1 = (:(epsilon_svr.epsilon), LogUniform(10e-9, 10e9))
r2 = (:(epsilon_svr.gamma), LogUniform(10e-9, 10e9))
r3 = (:(pca.pratio), [0.75, 0.9, 0.95, 0.99])
r4 = (:(feature_selector.features), [
    name -> false,
    name -> occursin("48.3_-3.0", string(name)),
    name -> occursin("48.2_-3.0", string(name)),
    name -> occursin("48.3_-3.1", string(name)),
    name -> occursin("48.1_-3.0", string(name)),
    name -> occursin("48.2_-3.1", string(name)),
    name -> occursin("48.1_-3.1", string(name)),
    name -> occursin("48.1_-3.2", string(name)),
    name -> occursin("48.2_-3.2", string(name)),
    name -> occursin("48.2_-3.3", string(name))    
])

tuned_model = TunedModel(model = pipe,
                         resampling = Holdout(; fraction_train=0.75),
                         tuning = RandomSearch(),
                         range = [r1, r2, r3, r4],
                         measure = smape,
                         n = 100,
                         acceleration = CPUProcesses())
mach_tuned_model = machine(tuned_model, X, y)

tuned_model.n = 4
fit!(mach_tuned_model)

tuned_model.n = 254
start = now()
fit!(mach_tuned_model)
stop = now()
mach_tuned_model.report.best_model

println("BEST SMAPE IS : " * string(mach_tuned_model.report.best_history_entry.measurement[1]))
println("TOTAL TIME IS : " * string((stop - start).value/1000) * "s")
# 31.5%, 435 seconds

MLJ.save("my_machine.jlso", mach_tuned_model)

File: summary.py
import pandas, os, time
import sklearn
from sklearn.svm import SVR
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.feature_selection import SelectorMixin
from sklearn.base import BaseEstimator
from sklearn.pipeline import make_pipeline
from sklearn.compose import TransformedTargetRegressor
from sklearn.model_selection import RandomizedSearchCV
from sklearn.model_selection import PredefinedSplit
from sklearn.metrics import make_scorer

import scipy.stats as stats
from joblib import dump, load

from numpy.random import seed
import random
SEED = 90125
seed(SEED)

path = os.path.join(os.environ["VOLTAIRE"], "data", "datasets", "SILFIAC")
data = pandas.read_csv(os.path.join(path, "train.csv"))

labels = [f"power_{x}" for x in range(24)]
y = data.loc[:, labels].mean(axis=1)
X = data.drop(columns = np.concatenate((np.array(labels), np.array(["date_col"]))))

class FeatureSelector(SelectorMixin, BaseEstimator):
    def __init__(self, filter_):
        SelectorMixin.__init__(self)
        self.filter_ = filter_

    def _get_support_mask(self):
        return np.array([self.filter_(name) for name in self.names])

    def fit(self, X, y=None):
        self.names = X.keys()
        return self

pipe = make_pipeline(
    FeatureSelector(lambda x: "48.3_-3.0" not in x),
    StandardScaler(),
    PCA(),
    SVR(max_iter=500000))

regr = TransformedTargetRegressor(
    regressor=pipe,
    func = lambda x: np.log(x+1),
    inverse_func = lambda x: np.exp(x)-1)
regr.fit(X, y)
regr.predict(X)

def smape(y_true, y_pred):
    return 200 * np.mean(np.mean(
        np.abs(y_pred - y_true) / (np.abs(y_true) + np.abs(y_pred) + np.finfo(np.float64).eps)))
smape(y, regr.predict(X))

mask = -np.ones(X.shape[0])
mask[np.array([d[:4] == "2019" for d in data.date_col])] = 0
cv = PredefinedSplit(mask)

distributions = {"regressor__svr__C" : stats.loguniform(10e-9, 10e9),
                 "regressor__svr__gamma" : stats.loguniform(10e-9, 10e9),
                 "regressor__pca__n_components" : [0.75, 0.9, 0.95, 0.99],
                 "regressor__featureselector__filter_" : [
                     lambda x : True,
                     lambda x : "48.3_-3.0" not in x,
                     lambda x : "48.2_-3.0" not in x,
                     lambda x : "48.3_-3.1" not in x,
                     lambda x : "48.1_-3.0" not in x,
                     lambda x : "48.2_-3.1" not in x,
                     lambda x : "48.1_-3.1" not in x,
                     lambda x : "48.1_-3.2" not in x,
                     lambda x : "48.2_-3.2" not in x,
                     lambda x : "48.2_-3.3" not in x]}
grid = RandomizedSearchCV(
    regr, distributions,
    scoring=make_scorer(smape, greater_is_better=False),
    n_jobs=-1, cv=cv, n_iter=250)
start = time.time()
grid.fit(X, y)
stop = time.time()

print("BEST SMAPE IS : ", -grid.cv_results_["mean_test_score"][np.logical_not(np.isnan(grid.cv_results_["mean_test_score"]))].max())
print("TOTAL TIME IS : ", stop-start, "s")
# 33.44%, 182s

dump(grid, 'my_grid.joblib')
